samcli/lib/bootstrap/companion_stack/companion_stack_manager.py

```diff
@@ -0,0 +1,319 @@
+            for _, current_repo in current_mapping.items():
+                if current_repo.logical_id == deployed_repo.logical_id:
+                    break
+            else:
+                unreferenced_repos.append(deployed_repo)
+        return unreferenced_repos
+
+    def delete_unreferenced_repos(self) -> None:
+        """
+        Blocking call to delete all deployed ECR repos that are unreferenced by a function
+        If repo does not exist, this will simply skip it.
+        """
+        repos = self.get_unreferenced_repos()
+        for repo in repos:
+            try:
+                self._ecr_client.delete_repository(repositoryName=repo.physical_id, force=True)
+            except self._ecr_client.exceptions.RepositoryNotFoundException:
+                LOG.debug("Image repo [%s] not found in companion stack. Skipping deletion.", repo.physical_id)
+
+    def sync_repos(self) -> None:
+        """
+        Blocking call to sync companion stack with the following actions
+        Creates the stack if it does not exist, and updates it if it does.
+        Deletes unreferenced repos if they exist.
+        Deletes companion stack if there isn't any repo left.
+        """
+        has_repo = bool(self.get_repository_mapping())
+        if self.does_companion_stack_exist():
+            self.delete_unreferenced_repos()
+            if has_repo:
+                self.update_companion_stack()
+            else:
+                self.delete_companion_stack()
+        elif has_repo:
+            self.update_companion_stack()
+
+    def does_companion_stack_exist(self) -> bool:
+        """
+        Does companion stack exist
+
+        Returns
+        -------
+        bool
+            Returns True if companion stack exists
+        """
+        try:
+            self._cfn_client.describe_stacks(StackName=self._companion_stack.stack_name)
+            return True
+        except ClientError as e:
+            error_message = e.response.get("Error", {}).get("Message")
+            if error_message == f"Stack with id {self._companion_stack.stack_name} does not exist":
+                return False
+            raise e
+
+    def get_repository_mapping(self) -> Dict[str, str]:
+        """
+        Get current function to repo mapping
+
+        Returns
+        -------
+        Dict[str, str]
+            Dictionary with key as function logical ID and value as ECR repo URI.
+        """
+        return dict((k, self.get_repo_uri(v)) for (k, v) in self._builder.repo_mapping.items())
+
+    def get_repo_uri(self, repo: ECRRepo) -> str:
+        """
+        Get repo URI for a ECR repo
+
+        Parameters
+        ----------
+        repo: ECRRepo
+
+        Returns
+        -------
+        str
+            ECR repo URI based on account ID and region.
+        """
+        return repo.get_repo_uri(self._account_id, self._region_name)
+
+    def is_repo_uri(self, repo_uri: Optional[str], function_logical_id: str) -> bool:
```

Comment:
Since these are methods of CompanionStackManager class, this context should be enough for it to not be used for generic ECR repo URIs.
My concern on adding companion stack to the method name is that we will need to add this for all other methods referring repos for consistency which can be too verbose.
```
comapnion_stack_manager.is_companion_repo_uri()
manager.get_companion_repos()
manager.get_unreferenced_companion_repos()
```